### PR TITLE
[wip] V2.0 refactor

### PR DESCRIPTION
**Pull request by rdelval**
Following Bill Kennedy's go workshop, I've decided to undertake a rewrite of this library to simplify the API and to get rid of unnecessary abstractions (i.e. interfaces).

Multiple other things will be changed, hope to make code changes as little painful as possible but I'm taking the time to create this refactor as a clean break which may incur immediate pain now and (hopefully) lead to much better maintenance.

Support for context will be added in this change which will also see us upgrading to Thrift 0.12.0 if all goes well.

### CODE REVIEW

File: realis.go [hunk 1]
```diff
@@ -256,11 +201,11 @@ func newTBinTransport(url string, timeout int, config *RealisConfig) (thrift.TTr
 // It will retry all calls which result in a temporary failure as well as calls that fail due to an EOF
 // being returned by the http client. Most permanent failures are now being caught by the thriftCallWithRetries
 // function and not being retried but there may be corner cases not yet handled.
-func NewRealisClient(options ...ClientOption) (Realis, error) {
+func NewRealisClient(options ...ClientOption) (*RealisClient, error) {
 	config := &RealisConfig{}
 
 	// Default configs
-	config.timeoutms = 10000
+	config.timeout = 10 * time.Second
```
Comment (at15) [inline realis.go]: better make all the default timers constant so you can track where they are used

File: updatejob.go [hunk 2]
```diff
@@ -140,7 +140,7 @@ func (u *UpdateJob) RollbackOnFail(rollback bool) *UpdateJob {
 
 func NewUpdateSettings() *aurora.JobUpdateSettings {
 
-	us := new(aurora.JobUpdateSettings)
+	us := aurora.JobUpdateSettings{}
```
Comment (at15) [inline updatejob.go]: struct literal

**Comment by rdelval** [comment]
@at15 Context, I'm working on as the next piece of the puzzle. I'm going to pull in and compile Thrift 0.12.0 and make sure everything works as expected (Thrift 0.11.0+ has context support). Will address the rest of your feedback individually as I get to it. Thanks for the review!

File: monitors.go [hunk 3]
```diff
@@ -19,49 +19,47 @@ import (
...
 )
 
 const (
 	UpdateFailed = "update failed"
 	RolledBack   = "update rolled back"
-	Timeout      = "timeout"
+	Timedout     = "timeout"
 )
 
 type Monitor struct {
```
[the remark this answers is not in the document]
Comment (rdelval) [reply monitors.go]: I agree, the way to create a monitor is very cumbersome. I like the Monitor* idea now that interfaces are gone in V2. I'll try to come up with other solutions but that seems like the less surprising one.

File: job.go [hunk 4]
```diff
@@ -20,57 +20,20 @@ import (
...
 	jobConfig *aurora.JobConfiguration
 	resources map[string]*aurora.Resource
 	portCount int
 }
 
-// Create a Job object with everything initialized.
-func NewJob() Job {
+// Create a AuroraJob object with everything initialized.
+func NewJob() *AuroraJob {
 	jobConfig := aurora.NewJobConfiguration()
```
[the remark this answers is not in the document]
Comment (rdelval) [reply job.go]: Done: https://github.com/rdelval/gorealis/commit/7967270b3b974fdcc107b0a40e7bfc8ad17325ad

File: examples/client.go [hunk 5]
```diff
@@ -243,19 +243,17 @@ func main() {
 
 	case "startCron":
 		fmt.Println("Starting a Cron job")
-		resp, err := r.StartCronJob(job.JobKey())
+		err := r.StartCronJob(job.JobKey())
```
Comment (at15) [inline examples/client.go]: might just have `fatalOnErr(err error)` and become onliner `fatalOnErr(r.StartCronJob(job.JobKey()))`, it makes example shorter and harder to copy and paste, which force user to think twice

````go
func fatalOnErr(err error) {
   if err != nil {
        log.Fatal(err)
   }
}
````

**Comment by at15** [comment]
@rdelval sorry for the late reply :sweat_smile:  for update job, I think the most important thing is document it's behavior and make the API self document

## Document

Behavior on rolling forward

- when to use job update over create job, add instances, kill instances (I guess update is aurora doing the combination on its own so client can simpler CRUD and monitor logic?)
- what you can do with the job update API
  - create job, it's like upsert, if no existing job, it create a new one
  - make a job running different tasks by updating an existing job
- when you change the instance count in job update, regardless of the id range for updates, it's a flex down
- the meaning of the settings, i.e. max instance failure, wait instance running time, batch size etc. and their impact on the update speed
- pause update etc. (I never use it, but I see people used it, in commented out code with don't delete prefix)

Behavior on invalid input

- if I give invalid id range, what would happen, aurora will reject or update everything or pick the valid part (if any)
- if I give extremely large instance count that the cluster can't fullfill, it will pending and never roll forwarded? (I am not sure about this ...)
- if the update have invalid constraint for schedule, pending (this is what I see)

Behavior on abort & rollback

- with autorollback, user don't need to do a extra new update to revert a failed one because aurora already did it
- when can you rollback, it seems only when rolling forward it will work
- when can you abort, what happens after abort, can you rollback aborted update
- what happens when there are multiple updates on same job

## API

Reduce footprint of thrift generated code, ideally, a gorealis user should not import `aurora` package at all to deal with all the `GetGetXXX_().GetUpdateXXX_()` it's like writing javascript and checking undefined `if a && a.b && a.b.c && a.b.c.d`  But this works might be too much ...

- id range should be either `map[int]bool` or `[]int`
- update key can have methods with value receivers like `func (u UpdateKey) Abort(cli *RealisClient)` `Rollback(cli *RealisClient)` so user know what they can do with a job update without loop through all the methods 

A pseduo code would be (sorry for the poor indent ... )

- all construct update using struct literal, `settings.A =1, settings.B = 2` makes the code hard to read
  - gorealis might need to do some sanity check and merge default values
- put method on the update key so user can know what is supported on update (editor will have hint on all the struct methods), and add document to these struct methods so jump to method gives them everything
- a `Wait` method to simplify monitor for lazy user

````
req := UpdateConfig{
     Job: {
         Env: a, 
        Meta: ....
    },
   Setting: {
       MaxFailurePerInstance: 3
   }
}

res, err := r.Update(req)
if err != nil {
    return err
}

// monitor manually to have better control
for {
select {
case ctx.Done():
    res.Abort("user cancel update")
case <-ticker.C: // poll ever 5s
     status, err := res.GetStatus() // replace the get update summary and filter
    if err != nil {
        return err
    }
    switch status.Code {
   case RolledForward:
        return nil
   case RolledBack:
       return errors.Errorf("rolled back: %s", status.String())
   case Rolling:
       // nothing
   default:
       return errors.Errorf("unknown status: %s %v", status.Code, status)
   }
}
}

// monitor implemented in gorealis for example or user that don't care about their app (like me)
finalStatus, err := res.Wait(ctx, 5 * time.Second) // block until it is finished or ctx is canceled
````

**Comment by rdelval** [comment]
Awesome feedback @at15. Documentation I've usually had little on the actual behavior because I don't want to duplicate already available documentation from the Aurora website but you do raise a good point that it would be nice to have the information be self contained within the project. I will make an effort to add unexpected behavior to the documentation.

Re: API, I'll add a helper struct Range:
```
type Range struct{
Start, End int
}
```

That should make things more manageable. 

For now re: updates, I've changed the anonymous struct to actually have a name and be accessible by the user. That way, a user can edit the job configuration passed in as well as the UpdateSettings. 

I'll have another update posted by tonight.

File: .travis.yml [hunk 6]
```diff
@@ -3,7 +3,7 @@ sudo: required
 language: go
 
 go:
-  - "1.10.x"
+  - "1.11.x"
```
Comment (at15) [inline .travis.yml]: so library using gorealis have to update to 1.11 now?

File: examples/client.go [hunk 7]
```diff
@@ -24,17 +24,16 @@ import (
...
 	"github.com/paypal/gorealis/gen-go/apache/aurora"
-	"github.com/paypal/gorealis/response"
 )
 
 var cmd, executor, url, clustersConfig, clusterName, updateId, username, password, zkUrl, hostList, role string
 var caCertsPath string
 var clientKey, clientCert string
 
-var CONNECTION_TIMEOUT = 20000
+var CONNECTION_TIMEOUT = 20 * time.Second
```
[the remark this answers is not in the document]
Comment (rdelval) [reply examples/client.go]: Done

File: updatejob.go (hunk 2 above)
Comment (rdelval) [reply updatejob.go]: Fixed with refactor

File: realis_e2e_test.go [hunk 8]
```diff
@@ -155,6 +155,50 @@ func TestLeaderFromZK(t *testing.T) {
...
+	assert.Error(t, err)
+	assert.Nil(t, r)
+
+	r, err = realis.NewClient(realis.SchedulerUrl("https://doesntexist.com:8081/testing/api"))
+	assert.Error(t, err)
+	assert.Nil(t, r)
+}
+
+func TestValidAuroraURL(t *testing.T) {
+	r, err := realis.NewClient(realis.SchedulerUrl("http://domain.com:8081/api"))
```
Comment (at15) [inline realis_e2e_test.go]: you can write a for loop, or table test if you need extra message

````go
urls := []string{"a", b"}
for _, u := range urls {
    r, err := realis.NewClient(u)
    assert ...
}
````

**Comment by rdelval** [comment]
We have released this initial version as branch 2.0. We will continue development there for continence of testing and will release a tag v2.0.0 once we feel confident the API has stabilized. Thanks to @at15 for the valuable constructive feedback along the way.

**Pull request closed without merging**